### PR TITLE
AMDGPU: Fix assembler asserting on expressions in vop3 instructions

### LINKED ISSUE #100075: [AMDGPU] Assembler cannot handle v_mad_u32_u24 with relocated operand
With this [test case](https://github.com/user-attachments/files/16345638/r.txt) I get:
```
$ llc -mtriple=amdgcn -mcpu=gfx1100 r.txt -o - | llvm-mc -triple=amdgcn -mcpu=gfx1100
	.text
	.section	.AMDGPU.config,"",@progbits
	.long	47176
	.long	1613692928
	.long	47180
	.long	0
	.long	47200
	.long	0
	.long	4
	.long	0
	.long	8
	.long	0
	.text
	.globl	f
	.p2align	8
	.type	f,@function
f:
	v_dual_mov_b32 v1, 0 :: v_dual_and_b32 v0, 1, v0
	s_delay_alu instid0(VALU_DEP_1)
unhandled operand type
UNREACHABLE executed at /home/jayfoad2/git/llvm-project/llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:8655!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
```
The instruction it fails on is `v_mad_u32_u24 v0, v0, 12, g0@abs32@lo`

### PR DESCRIPTION
**Pull request by arsenm**
Fixes #100075

### CODE REVIEW

**Comment by arsenm** [comment]
* **#100103** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100103?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by kosarev — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp [hunk 1]
```diff
@@ -8649,7 +8649,7 @@ void AMDGPUAsmParser::cvtVOP3(MCInst &Inst, const OperandVector &Operands,
       Op.addRegOrImmWithFPInputModsOperands(Inst, 2);
     } else if (Op.isImmModifier()) {
       OptionalIdx[Op.getImmTy()] = I;
-    } else if (Op.isRegOrImm()) {
+    } else if (Op.isRegOrImm() || Op.isExpr()) {
       Op.addRegOrImmOperands(Inst, 1);
     } else {
       llvm_unreachable("unhandled operand type");
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:8655]: Maybe don't bother checking the type here, just leave it to assert inside `addRegOrImmOperands`?
```suggestion
    } else {
      Op.addRegOrImmOperands(Inst, 1);
```

**Review by jayfoad — APPROVED** [review]
LGTM with or without the suggestion - thanks!

**Review by jayfoad — APPROVED** [review]

**Comment by jayfoad** [comment]
Is this worth backporting to 19.x?

**Comment by arsenm** [comment]
> Is this worth backporting to 19.x?

Sure

**Pull request merged**